server/src/main/java/org/elasticsearch/rest/action/document/RestBulkAction.java

```diff
@@ -62,6 +64,29 @@ public class RestBulkAction extends BaseRestHandler {
 
     public static final String TYPES_DEPRECATION_MESSAGE = "[types removal] Specifying types in bulk requests is deprecated.";
     public static final String FAILURE_STORE_STATUS_CAPABILITY = "failure_store_status";
+    public static final String PREFIX_LENGTH_FORMAT_CAPABILITY = "prefix_length_format";
+
+    private static final String BULK_FORMAT_HEADER = "X-Bulk-Format";
```

Comment:
I do not think we should introduce a separate header for this feature. Instead, we should let clients say that the request body is a length-prefixed bulk request using an appropriate value for the existing `Content-type` header. We already define a few vendor-specific content types with the `application/vnd.elasticsearch+` prefix. Let's come up with a way to add some more of them.

Reply:
Yes that's my thinking.

I'd suggest we stick with the `application/vnd.elasticsearch+` prefix since this is a vendor-specific format we're defining. I also don't think we need to fiddle with a separate `format=` parameter, we can leave `application/x-ndjson` as (effectively) the only delimited format and _define_ `application/vnd.elasticsearch+json.stream` etc as being length-prefixed.

> for json I think we should stick with ascii encoding

FWIW I disagree, I'd rather we used the same length-encoding scheme for all formats, and I'd rather it was a fixed number of bytes. Length-prefixed messages are awful to construct in a text editor regardless of the length encoding, you'd be kinda mad not to use some simple script to do this instead.

Reply:
Having Content-Type headers that do not resolve on XContentType adds an incredible amount of complexity in how rest request are handled. Not sure how to proceed, do you have any suggestion?

Reply:
First try by adding a XContentLengthPrefixedStreamingType in https://github.com/elastic/elasticsearch/pull/135506/commits/fc7513aaf1b291a1140615f6b08f626786075b16